app_integrations/apps/gsuite.py

```diff
@@ -0,0 +1,260 @@
+        self._activities_service = None
+        self._next_page_token = None
+
+    @classmethod
+    def _type(cls):
+        raise NotImplementedError('Subclasses should implement the _type method')
+
+    @classmethod
+    def service(cls):
+        return 'gsuite'
+
+    @classmethod
+    def date_formatter(cls):
+        """Return a format string for a date, ie: 2010-10-28T10:26:35.000Z"""
+        return '%Y-%m-%dT%H:%M:%SZ'
+
+    @classmethod
+    def _load_credentials(cls, keydata):
+        """Load ServiceAccountCredentials from Google service account JSON keyfile
+
+        Args:
+            keydata (dict): The loaded keyfile data from a Google service account
+                JSON file
+        """
+        try:
+            creds = ServiceAccountCredentials.from_json_keyfile_dict(
+                keydata, scopes=cls._SCOPES)
+        except (ValueError, KeyError):
+            # This has the potential to raise errors. See: https://tinyurl.com/y8q5e9rm
+            LOGGER.exception('Could not generate credentials from keyfile for %s',
+                             cls.type())
+            return False
+
+        return creds
+
+    def _create_service(self):
+        """GSuite requests must be signed with the keyfile
+
+        Returns:
+            bool: True if the Google API discovery service was successfully established or False
+                if any errors occurred during the creation of the Google discovery service,
+        """
+        if self._activities_service:
+            LOGGER.debug('Service already instantiated for %s', self.type())
+            return True
+
+        creds = self._load_credentials(self._config.auth['keyfile'])
+        if not creds:
+            return False
+
+        try:
+            resource = discovery.build('admin', 'reports_v1', credentials=creds)
+        except errors.Error:
+            LOGGER.exception('Failed to build discovery service for %s', self.type())
+            return False
+
+        # The google discovery service 'Resource' class that is returned by
+        # 'discovery.build' dynamically loads methods/attributes, so pylint will complain
+        # about no 'activities' member existing without the below pylint comment
+        self._activities_service = resource.activities() # pylint: disable=no-member
+
+        return True
+
+    def _gather_logs(self):
+        """Gather the G Suite Admin Report logs for this application type
+
+        Returns:
+            bool or list: If the execution fails for some reason, return False.
+                Otherwise, return a list of activies for this application type.
+        """
+        if not self._create_service():
+            return False
+
+        activities_list = self._activities_service.list(
+            userKey='all',
+            applicationName=self._type(),
+            startTime=self._last_timestamp,
+            pageToken=self._next_page_token if self._next_page_token else None
```

Comment:
Is there some reason that `self._next_page_token` could be an empty string? If so, you could do simply:

```python
pageToken = self._next_page_token or None
```

Reply:
The case of it potentially being an empty string is exactly why I added this, so good catch :). I don't know for sure if passing `''` to `pageToken` would have a different effect than passing `None` so I just wanted to guard against it.

I added this because I'm not sure what the default value (if any) is within the [response](https://developers.google.com/admin-sdk/reports/v1/reference/activities/list) for `nextPageToken` that gets returned. It's unclear if 1) the key exists at all if there is no value, and 2) what the value would be if it _does_ exist but is empty (is it an empty string or a json null)?

The code where this is set is here, using a `.get`:
https://github.com/airbnb/streamalert/blob/14480665dc214185f6c69986abf8092ddb3badb8/app_integrations/apps/gsuite.py#L121

Reply:
Maybe a second set of eyes on the docs would help unearth something I'm not seeing ;)